tests/system/robot/NVDAInstaller.robot

```diff
@@ -0,0 +1,43 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2021 NV Access Limited
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+*** Settings ***
+Documentation	Smoke test the installation process
+Force Tags	NVDA	smoke test	installer
+
+# for start & quit in Test Setup and Test Teardown
+Library	NvdaLib.py
+Library	NVDAInstaller.py
+Library	ScreenCapLibrary
+
+Test Setup	default startup
+Test Teardown	default teardown
+
+*** Keywords ***
+default exit if no installDir
+	Pass Execution If	not r'${installDir}'	"skipping due to not running on installer"
+
+default teardown
+	default exit if no installDir
+	${screenshotName}=	create_preserved_test_output_filename	failedTest.png
+	Run Keyword If Test Failed	Take Screenshot	${screenShotName}
+	quit NVDAInstaller
+
+default startup
+	default exit if no installDir
+	start NVDAInstaller	standard-dontShowWelcomeDialog.ini
+
+default pass execution
+
+*** Test Cases ***
+
+Read install dialog
+	default exit if no installDir
```

Comment:
Because this is in the default startup I don't think it needs to be in the test?

Reply:
weirdly, `Pass Execution If` skips the rest of the current setup, teardown and test, meaning it has to be done in each. Spent a while trying to find a simpler fix after noticing the behaviour. 

https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Pass%20Execution%20If

Reply:
Does this mean that it skips the rest of the setup / startup step, but then tries to continue with the test?

If so, [the docs for pass execution](https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Pass%20Execution) seem to suggest that tags can be added. Perhaps, the `Pass Execution If` could add or remove tags that prevent the tests that depend on `installDir` from running?

Reply:
>Does this mean that it skips the rest of the setup / startup step, but then tries to continue with the test?

Correct

> Perhaps, the `Pass Execution If `could add or remove tags that prevent the tests that depend on installDir from running?

I tested this with the following code, by adding the `excluded_from_build` tag and removing the `installer` tag. The changes were logged properly but the tests were triggered (hence the changes logged). I think the choice of tests are done at the start of the robot test process. 

```
*** Keywords ***
default teardown
	${screenshotName}=	create_preserved_test_output_filename	failedTest.png
	Run Keyword If Test Failed	Take Screenshot	${screenShotName}
	quit NVDAInstaller

default startup
	Pass Execution If	not r'${installDir}'	"skipping due to not running on installer"	excluded_from_build	-installer
	start NVDAInstaller	standard-dontShowWelcomeDialog.ini

default pass execution

*** Test Cases ***

Read install dialog
	Log To Console	"@{TEST TAGS}"
	[Documentation]	Ensure that the install dialog can be read in full
	read_install_dialog	# run test

Read install dialog portable copy
	Log To Console	"@{TEST TAGS}"
	[Documentation]	Ensure that the portable copy install dialog can be read in full
	read_portable_copy_dialog	# run test

```

Reply:
Yes, it looks like it isn't possible to change the tests dynamically: https://stackoverflow.com/questions/26369374/how-to-skip-a-specific-testcase-in-robot-framework

You might be able to configure this such that the installer tests are not in the set of tests that are run by default. Eg. a tag must be supplied by the robot invocation in order to make it run.

Reply:
Then it would be an error for the suit if the `installDir` argument is not supplied. The suite startup is only run once, and is different from the test startup, which you have in the previous example. For an example of how to configure it see https://stackoverflow.com/questions/45204684/robot-framework-differences-between-suite-setup-and-test-setup. From there you could use `fail` rather than `pass execution if`

Reply:
I from what I could tell, there was no way to override a variable that had already been passed in `eg --excluded exclude_from_build --include installer` doesn't work, so AppVeyor wouln't be able to use runsystemtests.bat. 

Maybe this is worth doing though and just putting the manual command in AppVeyor without a new batch file and robotArgs file

Reply:
I suspect that if the tags to include are specified when robot is first run, it will only run tests matching that tag. In that case we can make the "default set of tests" only the tests with the "NVDA" tag. The installer tests wouldn't have this tag, instead they would have an `installer` tag. Then in the `robotArgs.robot` file (which has all the default options for running Robot framework), add a line to include tests with the `NVDA` tag.

Then when running from appveyor, we also include the `installer` tag.

Reply:
I may have misunderstood you. Do you mean that when running robot from the command line, if an option is specified in the `robotArgs.robot` file, the same argument can't be specified as a command line argument as well?

Reply:
It looks like it should be possible to at least override the --include argument see
From: http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#using-argument-files

> When an argument file is used with other arguments, its contents are placed into the original list of arguments to the same place where the argument file option was. This means that options in argument files can override options before it, and its options can be overridden by options after it. It is possible to use --argumentfile option multiple times or even recursively

I've just tested this locally and it seems to work. Using `--dryrun` makes it faster to see which tests will run.

Reply:
To be clear, specifying `--include` on the command line in addition to in the `rogotArgs.robot` file resulted in additive behavior. Tests with both tags were included.

Reply:
Thanks for looking into this, I figured there would be an easy option but as you would have noticed, the robot framework is somewhat unintuitive.